Add optional SELinux support

flyn-org opened the pull request:
These patches capture the work of Thomas Petazzoni to add optional SELinux support to OpenWrt. He first proposed this work on the openwrt-mailing list:

http://lists.infradead.org/pipermail/openwrt-devel/2019-November/025973.html
http://lists.infradead.org/pipermail/openwrt-devel/2019-November/025974.html
http://lists.infradead.org/pipermail/openwrt-devel/2019-November/025975.html
http://lists.infradead.org/pipermail/openwrt-devel/2019-November/025976.html
http://lists.infradead.org/pipermail/openwrt-devel/2019-November/025977.html
http://lists.infradead.org/pipermail/openwrt-devel/2019-November/025978.html
http://lists.infradead.org/pipermail/openwrt-devel/2019-November/025979.html
http://lists.infradead.org/pipermail/openwrt-devel/2019-November/025980.html

I have adapted them to a more current upstream master.

The corresponding package work by Thomas is available at https://github.com/openwrt/packages/pull/10664.

adrianschmutzler commented:
Drive by comment, but please make sure you have proper commit title prefixes. At least the first commit lacks a commit message.

Despite that, it appears that you have changed Thomas' patches and neither marked the changes as yours nor added your Signed-off-by anywhere?

flyn-org commented:
@tpetazzoni, is there anything special I need to do to boot with SELinux active? I tried a vanilla boot, and I also tried booting with `selinux=1` and `enforcing=1` passed in the kernel boot options. In any case, `setenforcing 1` prints `setenforce: SELinux is disabled`.

`dmesg` prints `LSM: Security Framework initializing`, but nothing about SELinux.

From what I can tell, the necessary SELinux statements exist in the kernel's `.config`, but it seems as if the kernel does not contain SELinux support.

tpetazzoni commented:
I was testing under Qemu, and my Qemu script was passing security=selinux to select selinux as the LSM.

flyn-org commented:
I am using QEMU to test too. Passing `security=selinux` does indeed seem to activate SELinux. I hadn't come across the `security=` parameter yet (which seems to select the LSM model). I am pretty familiar with things from this point on. Thanks!

adrianschmutzler commented:
@flyn-org Please take care of properly marking authorship as pointed out or I will have to close this.

flyn-org commented:
@adrianschmutzler, I  kept @tpetazzoni's name in place because he was the original author or these patches. I did not do anything other than grab the patches from the openwrt-devel mailing list, apply them to Git, and push them here. I want the credit to remain with the original author.

Should I leave Thomas as the author, but change the signed-off-by statement to my name?

adrianschmutzler commented:
@flyn-org But you didn't take the patches unchanged. From a quick look I can tell that you at least changed the commit title. That (and all other changes compared to Thomas' version should be marked by _adding_ your Signed-off-by at the end of the commit message, e.g.

```
busybox: add optional selinux support

blabla commit message

Signed-off-by: Thomas Petazzoni <thomas.petazzoni@bootlin.com>
[add commit message, adjust commit title]
Signed-off-by: You
```

This will allow to identify what's from Thomas and where you altered his work (and made it better - or worse). Keeping the original Author and Signed-off-by of course is actually a good thing; you just have to mark your changes, unless you keep the thing completely untouched.

Despite that, note that most of the commit still need some work regarding the commit title (proper prefix and overall length).

bobafetthotmail commented:
You said you refreshed the patches so you can add something like
"patches refreshed for current master by"
Signed-off-by (your name) (your email)"

after his Signed-off-by.

tpetazzoni committed:
build: add support for SELinux to include/image.mk

This allows the build process to prepare a squashfs filesystem for use
with SELinux.

Signed-off-by: Thomas Petazzoni <thomas.petazzoni@bootlin.com>
[rebase, add commit message]
Signed-off-by: W. Michael Petullo <mike@flyn.org>

tpetazzoni committed:
tools/squashfskit4: add xattr support to mksquashfs for SELinux

SELinux uses extended attributes to store SELinux security contexts.

Signed-off-by: Thomas Petazzoni <thomas.petazzoni@bootlin.com>
[rebase, add commit message]
Signed-off-by: W. Michael Petullo <mike@flyn.org>

flyn-org commented:
@adrianschmutzler and @bobafetthotmail, I adjusted the commit messages in a way that preserves Thomas' contribution while meeting your intent.

Things are very close. I still need to fix the Python package at https://github.com/openwrt/packages/pull/10664. Other than that, things seem to work. Here is how I am testing:

1. Run `git clone git@github.com:flyn-org/openwrt -b selinux`.

2. Set feeds.conf to make use of:

        src-git packages git@github.com:tpetazzoni/openwrt-packages.git;selinux

3. Run `./scripts/feeds update`.

4. Run `./scripts/feeds install python3 pcre libselinux audit libcap-ng policycoreutils libsemanage checkpolicy refpolicy selinux-python`.

5. Run `make menuconfig` and select:

  * x86_64 build
  * Global build settings->Kernel build options->Ext4 Security Labels
  * Base system->busybox->Customize busybox options
  * Settings->Support NSA Security Enhanced Linux
  * Archival utilities->Support extracting SELinux labels
  * SELinux Utilities->[Select all]
  * Languages->Python->python-sepolgen and python-sepolicy
  * Utilities->selinux-*
  * Administration->refpolicy

6. Run `make V=99`.

7. Extract `openwrt-x86-64-generic-ext4-combined.img` and run it with
`qemu-system-x86_64 -hda openwrt-x86-64-generic-ext4-combined.img -net
nic -net user`. When prompted by GRUB, add `security=selinux` to the
kernel's boot parameters.

8. Confirm the kernel loaded SELinux by running `sestatus`.

9. Label the filesystem by running `restorecon -r /`.

10. Set enforcing mode by running `setenforce 1`.

flyn-org commented:
I just fixed the Python package. Everything should be ready to test.

tpetazzoni commented:
Why do you need the filesystem labelling ysing restorecon -r / at boot time ? At least with my patches, if you're using a SquashFS filesystem, the security context of the files/directories are set at build time, using setfiles. I think the same should be done for other filesystems: the image produced by OpenWRT should be directly bootable in enforcing mode, without the need for a first "dummy" boot.

adrianschmutzler commented:
Just had a look at our patchwork to tidy up stuff there. As it appears, the SELinux patchset, submitted in November 2019, has essentially remained mostly uncommented and unreviewed despite some minimal feedback since then. For this PR, open for already 15 days now, situation does not appear to be different, just with added formal issues on top.

Therefore, it seems quite obvious to me that this situation won't change in the near future. So, to save us all from wasted time and efforts, I'm going to crush your remaining hopes and close this now. Please don't feel repelled, and continue to contribute to OpenWrt in the future, hopefully with more luck next time!

aparcar commented on tools/fakeroot/Makefile:
```diff
@@ -0,0 +1,20 @@
+PKG_VERSION:=1.20.2
+
+PKG_SOURCE:=$(PKG_NAME)_$(PKG_VERSION).orig.tar.bz2
+PKG_SOURCE_URL:=http://snapshot.debian.org/archive/debian/20141005T221953Z/pool/main/f/fakeroot
```
Is there a less fishy source for fakeroot available?

tpetazzoni replied:
We could probably use https://salsa.debian.org/clint/fakeroot instead, which is what is referred to by Debian as being the Git repo for this package.

aparcar replied:
I would prefer that mirror as it's easier to update to future version. Updating the timestamp within the source URL seems a bad solution.

tpetazzoni replied:
Note that the complete source URL also includes the fakeroot version, i.e 1.20.2. The timestamp is just because snapshot.debian.org is archiving all Debian packages ever released in a timestamp based URL format.

flyn-org replied:
The GitLab repository at https://salsa.debian.org/clint/fakeroot does not seem to consistently tag releases, unless I am missing something. The version used here, 1.20.2, does not appear as a tag. This might make relying on the GitLab instance troublesome.

@tpetazzoni, have you seen tarballs archived at https://salsa.debian.org/clint/fakeroot? I could not find any.

Fedora uses https://tracker.debian.org/pkg/fakeroot as the URL for their packages, which is kind of odd.

Another option might be http://ftp.debian.org/debian/pool/main/f/fakeroot/. This is what good old freshmeat.net links to. Alpine uses http://ftp.debian.org/debian/pool/main/f/fakeroot/fakeroot_$pkgver.orig.tar.gz.

aparcar replied:
Anything where we just have to adapt the PKG_VERSION is fine for me. Based on the git a newer version is available (not tagged). Are any of the provided sources using the latest version?

tpetazzoni replied:
> The GitLab repository at https://salsa.debian.org/clint/fakeroot does not seem to consistently tag releases, unless I am missing something. The version used here, 1.20.2, does not appear as a tag. This might make relying on the GitLab instance troublesome.
> 
> @tpetazzoni, have you seen tarballs archived at https://salsa.debian.org/clint/fakeroot? I could not find any.
> 
> Fedora uses https://tracker.debian.org/pkg/fakeroot as the URL for their packages, which is kind of odd.
> 
> Another option might be http://ftp.debian.org/debian/pool/main/f/fakeroot/. This is what good old freshmeat.net links to. Alpine uses http://ftp.debian.org/debian/pool/main/f/fakeroot/fakeroot_$pkgver.orig.tar.gz.

The problem with http://ftp.debian.org/debian/pool/main/f/fakeroot/ URLs is that they are not stable. Once the version is no longer used in any Debian version, the tarball disappears from there. That's why in Buildroot we use snapshot.debian.net, because URLs are guaranteed to be stable.

aparcar commented on tools/squashfskit4/Makefile:
```diff
@@ -22,7 +22,7 @@ define Host/Compile
 		XZ_SUPPORT=1 \
 		LZMA_XZ_SUPPORT=1 \
-		XATTR_SUPPORT= \
+		XATTR_SUPPORT=1 \
```
This doesn't look optional, did you check the size changes?

tpetazzoni replied:
This additional feature is inside a host tool, it doesn't go on the target. Knowing that, is a size increase really a big problem here ?

aparcar replied:
No that shouldn't be a problem, thank you for the clarification.

aparcar commented:
I'm new to SELinux and would like to understand the concept. As I'm quite familiar with the OpenWrt build system I see myself qualified to review some chunks of this complex pull request. I'd therefore kindly ask to reopen this issue.

The reviewing may include minor fixes, however I don't see myself in the position to adopt SELinux as my personal project, therefore only do further testing if @tpetazzoni and/or @flyn-org  are up to further work on this.

adrianschmutzler commented:
I'm reopening this now, as at least individual people expressed interest.

This might be a good opportunity for the downvoters of my closure to start taking part in the discussion and development.

Apart from that, we still have the issue with proper Signing-off for most of the commits here. So, please don't just refer to some old patchwork patch, but properly implement authorship in the patches posted here.

adrianschmutzler commented:
@flyn-org This is a good example: "build: add support for SELinux to include/image.mk"

This patch is essentially taken unaltered from the patchwork page. You're only adding the commit message, so your commit message could look like:

```
build: add support for SELinux to include/image.mk

This allows the build process to prepare a squashfs filesystem for use
with SELinux.

Signed-off-by: Thomas Petazzoni <thomas.petazzoni@bootlin.com>
[rebase, add commit message]
Signed-off-by: W. Michael Petullo <mike@flyn.org>
```

One could even drop the "rebase", as in this case it might have been trivial.

Despite, I'm not sure whether our policy allows to abbreviate your first name in Signed-off-by.

aparcar commented:
@tpetazzoni could you please share your qemu test script?

tpetazzoni commented:
> @tpetazzoni could you please share your qemu test script?

I was just booting with

```
qemu-system-arm -M virt -cpu cortex-a15 -kernel openwrt-armvirt-32-zImage -append "console=ttyAMA0,115200 root=/dev/vda security=selinux" -serial stdio -drive file=openwrt-armvirt-32-rootfs-squashfs.img,if=none,format=raw,id=hd -device virtio-blk-device,drive=hd
```

as far as I can see from my notes.

aparcar commented:
Okay, an initial compile and run test:

[Bootlog](http://sprunge.us/6hyKlE)

```
root@OpenWrt:/# getenforce 
Permissive
root@OpenWrt:/# sestatus 
SELinux status:                 enabled
SELinuxfs mount:                /sys/fs/selinux
Current mode:                   permissive
Mode from config file:          permissive
Policy version:                 31
Policy from config file:        targeted
```

I started the VM via the following command. The `PARTUUID` stays the same if no source/kernel changes are done. I just copied it by running without external Kernel and copied it from the group entry.

```
qemu-system-x86_64 \
  -hda openwrt-x86-64-generic-squashfs-combined.img \
  -kernel openwrt-x86-64-generic-kernel.bin \
  -append "root=PARTUUID=8953a906-02 rootwait console=tty0 console=ttyS0,115200n8 noinitrd security=selinux" \
  -nographic 
```

So apart from the two comments I left at the packages.git PR, it seems to compile and run.

Can you give me an example to check it's actually doing something?

aparcar commented:
Could you please try to compile on the latest Kernel version? I'm getting plenty of questions on which kernel options to enable.

aparcar commented:
Labels seem to work fine on squashfs

```
system_u:object_r:bin_t          bin
system_u:object_r:unlabeled_t    boot
system_u:object_r:tmpfs_t        dev
system_u:object_r:etc_t          etc
system_u:object_r:default_t      init
system_u:object_r:lib_t          lib
system_u:object_r:lib_t          lib64
system_u:object_r:mnt_t          mnt
system_u:object_r:default_t      overlay
system_u:object_r:proc_t         proc
system_u:object_r:root_t         rom
root:object_r:user_home_dir_t    root
system_u:object_r:bin_t          sbin
system_u:object_r:sysfs_t        sys
system_u:object_r:tmpfs_t        tmp
system_u:object_r:usr_t          usr
system_u:object_r:default_t      var
system_u:object_r:default_t      www
```

flyn-org commented:
> @flyn-org This is a good example: "build: add support for SELinux to include/image.mk"
[...]
> Despite, I'm not sure whether our policy allows to abbreviate your first name in Signed-off-by.

Okay, now I have found some earlier commits that follow this format (i.e., the use of [...]). I will update this to use that style.

W. Michael Petullo is the name I use for all formal correspondence. It would seem odd if this were unacceptable, since OpenWrt collaborators span the globe, and there exist wildly different conventions on human names. I have used my name in previous OpenWrt commits; please see:
```
commit 8c7d6c47a7893db689e50e1917479ad07a96547a
Author: W. Michael Petullo <mike@flyn.org>
Date:   Sat Jun 29 20:06:20 2019 -0400
```
or
```
commit 92aa21497b22bcf03862cf9da8e89ef007affebd
Author: W. Michael Petullo <mike@flyn.org>
Date:   Sun Aug 26 02:31:31 2018 -0400
```

aparcar commented:
> Despite, I'm not sure whether our policy allows to abbreviate your first name in Signed-off-by.

@flyn-org I think that's fine based on the lack of a more explicit policy on [kernel.org](https://www.kernel.org/doc/html/v4.17/process/submitting-patches.html).

Please follow the kernel.org suggestions and use a line like @adrianschmutzler wrote:

```
Signed-off-by: Thomas Petazzoni <thomas.petazzoni@bootlin.com>
[rebase, add commit message]
Signed-off-by: W. Michael Petullo <mike@flyn.org>
```
There is a policy to have a commit message, which is missing at https://github.com/openwrt/openwrt/pull/3207/commits/c2b8e72185633006663aa5f5a144865be39870cb. Even if it's reasonable simple please describe the context.

adrianschmutzler commented:
> W. Michael Petullo is the name I use for all formal correspondence. It would seem odd if this were unacceptable, since OpenWrt collaborators span the globe, and there exist wildly different conventions on human names. I have used my name in previous OpenWrt commits; please see:

guidelines just say "real name", and since this was accepted previously a you stated, this should be fine.

flyn-org commented:
@adrianschmutzler and @aparcar, I think my latest push addresses your requests regarding the commit log format. Would someone kindly advise how to keep the kernel build from prompting for configuration parameters? Everything is presently based on the latest OpenWrt master.

adrianschmutzler commented:
Formally this looks like what I expected now, thanks!

aparcar commented:
Can you try the following:

    make target/{clean,compile} V=s

This fixed it for me. There where some missing symbols recently...

aparcar commented:
@dangowrt I ran this patch through [multiple CI SDKs](https://gitlab.com/aparcar/openwrt/-/jobs/669276450) with disabled SELinux support and the size changes seem minimal (few Bytes if at all). Based on the code changes this makes sense. @flyn-org can you verify those size changes? Getting code into OpenWrt that increases size is often much more tedious.

```shell
Compare packages of x86/64/x86_64:
busybox procd
Checking installed size
Fetching latest package indexes...
Comparing package sizes...
Change 	Local 	Package
9B	50416B	procd
2B	228828B	busybox
```

flyn-org commented:
> @flyn-org can you verify those size changes? Getting code into OpenWrt that increases size is often much more tedious.

Here are the number of bytes consumes by the base packages build with OpenWrt vanilla 48a9d99a, 48a9d99a + SELinux patched (selected), and 48a9d99a + SELinux patches (not selected/same configuration as vanilla):
```
Vanilla SELinux (active)  SELinux (inactive)
705374  705329            705329   base/bnx2-firmware_20200619-1_x86_64.ipk
229622  236978            229583   base/busybox_1.31.1-1_x86_64.ipk
        116367                     base/ca-bundle_20200601-1_all.ipk
130677  129887            129887   base/dnsmasq_2.82-3_x86_64.ipk
103687  103665            103665   base/dropbear_2020.80-1_x86_64.ipk
176706  176691            176691   base/e2fsprogs_1.45.6-1_x86_64.ipk
51504   51505             51505    base/firewall_2020-07-25-e9b90dfa-1_x86_64.ipk
2269    2259              2259     base/getrandom_2019-12-31-0e34af14-4_x86_64.ipk
7083    7086              7086     base/jshn_2020-07-11-f4e9bf73-1_x86_64.ipk
10474   10471             10471    base/jsonfilter_2018-02-04-c7e938d6-1_x86_64.ipk
127574  127580            127580   base/libblkid1_2.35.1-2_x86_64.ipk
4446    4442              4442     base/libblobmsg-json_2020-07-11-f4e9bf73-1_x86_64.ipk
        24933                      base/libbz2-1.0_1.0.8-1_x86_64.ipk
4855    4851              4851     base/libcomerr0_1.45.6-1_x86_64.ipk
174471  174473            174473   base/libext2fs2_1.45.6-1_x86_64.ipk
39108   39117             39117    base/libf2fs6_1.13.0-1_x86_64.ipk
27659   27654             27654    base/libjson-c5_0.14-1_x86_64.ipk
6013    6002              6002     base/libjson-script_2020-07-11-f4e9bf73-1_x86_64.ipk
        178181                     base/libncurses6_6.2-1_x86_64.ipk
16124   16118             16118    base/libnl-tiny_2019-10-29-0219008c-1_x86_64.ipk
        1512550                    base/libopenssl1.1_1.1.1g-1_x86_64.ipk
85941   85955             85955    base/libsmartcols1_2.35.1-2_x86_64.ipk
9451    9454              9454     base/libss2_1.45.6-1_x86_64.ipk
22353   22354             22354    base/libubox20191228_2020-07-11-f4e9bf73-1_x86_64.ipk
11476   11472             11472    base/libubus20191227_2020-02-05-171469e3-1_x86_64.ipk
17727   17726             17726    base/libuci20130104_2020-04-24-ec8d3233-3_x86_64.ipk
11313   11314             11314    base/libuclient20160123_2020-06-17-c6609861-1_x86_64.ipk
12210   12214             12214    base/libuuid1_2.35.1-2_x86_64.ipk
10899   10894             10894    base/logd_2019-12-31-0e34af14-4_x86_64.ipk
14796   14793             14793    base/mkf2fs_1.13.0-1_x86_64.ipk
82408   82395             82395    base/netifd_2020-06-06-51e9fb81-1_x86_64.ipk
27003   26995             26995    base/odhcp6c_2020-03-28-f575351c-16_x86_64.ipk
39264   39258             39258    base/odhcpd-ipv6only_2020-06-21-5da52992-4_x86_64.ipk
2099    2095              2095     base/openwrt-keyring_2019-07-25-8080ef34-1_x86_64.ipk
74259   74266             74266    base/opkg_2020-05-07-f2166a89-1_x86_64.ipk
18498   20938             18498    base/Packages
5949    6809              5936     base/Packages.gz
23285   26237             23285    base/Packages.manifest
151     151               151      base/Packages.sig
46553   46551             46551    base/partx-utils_2.35.1-2_x86_64.ipk
126288  126276            126276   base/ppp_2.4.8.git-2020-05-25-2_x86_64.ipk
11170   11165             11165    base/ppp-mod-pppoe_2.4.8.git-2020-05-25-2_x86_64.ipk
51186   51485             51182    base/procd_2020-07-30-2ae5cbcf-1_x86_64.ipk
27896   27884             27884    base/r8169-firmware_20200619-1_x86_64.ipk
        6949                       base/terminfo_6.2-1_x86_64.ipk
16667   16674             16674    base/ubox_2019-12-31-0e34af14-4_x86_64.ipk
6697    6696              6696     base/ubus_2020-02-05-171469e3-1_x86_64.ipk
12857   12858             12858    base/ubusd_2020-02-05-171469e3-1_x86_64.ipk
7884    7888              7888     base/uci_2020-04-24-ec8d3233-3_x86_64.ipk
7849    7853              7853     base/uclient-fetch_2020-06-17-c6609861-1_x86_64.ipk
1555    1553              1553     base/urandom-seed_2_x86_64.ipk
5677    5672              5672     base/urngd_2020-01-21-c7f7b6b6-1_x86_64.ipk
12125   12121             12121    base/usign_2020-05-23-f1f65026-1_x86_64.ipk
```

aparcar commented:
> Here are the number of bytes consumed by the base packages built with OpenWrt vanilla 48a9d99, 48a9d99 + SELinux patches (selected), and 48a9d99 + SELinux patches (not selected/same configuration as vanilla):

So the patch series actually improves the filesize, curious...

I'll check the `kernel.bin` size changes later. Glad that it doesn't increase the Packages much if it's not activated.

flyn-org commented:
Here are the number bytes consumed by the OpenWrt vanilla 48a9d99, 48a9d99 + SELinux patches (selected), and 48a9d99 + SELinux patches (not selected/same configuration as vanilla) kernels:

4866720	vanilla/bin/targets/x86/64/openwrt-x86-64-generic-kernel.bin
4965024	selinux-on/bin/targets/x86/64/openwrt-x86-64-generic-kernel.bin
4862624	selinux-off/bin/targets/x86/64/openwrt-x86-64-generic-kernel.bin

The decrease in selinux-off is indeed odd, but I double checked that I have the labels right. I built everything on the same Fedora 32 host.

aparcar commented:
> The decrease in selinux-off is indeed odd, but I double checked that I have the labels right. I built everything on the same Fedora 32 host.

I'll do a test build as well again including the kernel sizes. Thanks

aparcar commented:
I tried to improve the OpenWrt package building via [this patch series](https://patchwork.ozlabs.org/project/openwrt/list/?series=194718), which includes the here proposed fakeroot patch. Having fakeroot as host tools is a good first step to decrease this PR size.

aparcar commented:
Felix pulled an updated and MacOS compatible version of `fakeroot` in his [staging tree](https://git.openwrt.org/?p=openwrt/staging/nbd.git;a=commit;h=7925c9d9222e9a8d772a8b71bbee1f729b243228), please remove the Fakeroot commit from this pull request. @tpetazzoni stays the original committer and we just added additional SoB lines.

aparcar commented on package/system/procd/patches/0001-initd-init-add-minimal-SELinux-policy-loading-suppor.patch:
```diff
@@ -0,0 +1,110 @@
+selinux_init_load_policy() from libselinux, and then re-execute init
+so that it runs with the SELinux policy in place and enforced.
+
+Signed-off-by: Thomas Petazzoni <thomas.petazzoni at bootlin.com>
```
@tpetazzoni I think this does not count as a valid email address as it breaks `git send-email` auto CCing. Is the usage of ` at ` rather than `@` important to you? I'd figure that most spam bots are smart enough to figure out that `<.+at.+.com>` is an email address.

aparcar replied:
I checked the Linux Kernel and it doesn't seem to be used for final SoB lines, however as I corrected Open**WRT** and thereby added myself as final SoB with a correct mail address, I'm figuring it's fine.

tpetazzoni replied:
I don't know what replaced @ by "at", but it's certainly not intentional, and probably something done by one of the mailing list software on the way. You can and should definitely use my real e-mail address with an @ sign. Thanks!

aparcar replied:
Bad news, `procd` now contains your *spam proof email*. We'll fix that for the other cases

aparcar commented:
I sent a [*fixed* version of the `procd` patch](https://patchwork.ozlabs.org/project/openwrt/list/?series=194834) to the mailing list. It should be part of procd.git directly instead of `openwrt.git`.

adrianschmutzler commented:
> I think this does not count as a valid email address as it breaks git send-email auto CCing. Is the usage of at rather than @ important to you? I'd figure that most spam bots are smart enough to figure out that <.+at.+.com> is an email address.

I typically only merge with a "proper" e-mail address. This field is supposed to contain an e-mail address, so one should put (a valid) one there. However, I'm not sure whether this is codified somewhere.

aparcar commented:
@flyn-org please remove the fakeroot and procd patches, they are either merged or about to be merged.

As `procd` now has a hard dependency on `libselinux` and `libsepol` please move both packages from your `packages.git` PR over here. For `libselinux`, only the C part without `libselinux-python`. That needs to be separated as Python is not part of `openwrt.git`. Meaning, please add `libselinux` here and `python3-libselinux` to `packages.git`.

I'm currently working on splitting up `procd` and `busybox` into two variant, with SELinux support and without. Ideally this allows in the long run to enable SELinux on a target base, e.g. enabling it for x86 in while ath79 may never use it.

flyn-org commented:
I removed the procd and fakeroot changes, as these have been merged elsewhere. I also moved the libselinux (without Python) and libsepol packages here from the packages repository. I am presently performing a test build, so I might make some more small changes today.

aparcar commented:
@flyn-org thanks for the quick rework. I'm about to finish the procd/busybox variants. On recompiling the Kernel I get a bunch of symbol questions, does the same happen to you?

flyn-org commented:
@aparcar, yes, I have been unable to get the kernel symbol prompts to go away. By the way, I just pushed some small changes to libselinux/python-libselinux here and at https://github.com/openwrt/packages/pull/10664 that followed from a build test.

flyn-org commented:
Will the BusyBox work be pulled out as with the procd modifications?

aparcar commented:
> Will the BusyBox work be pulled out as with the procd modifications?

Yes I'll create a PR including both updated Makefiles for busybox and procd

bobafetthotmail commented:
> @flyn-org thanks for the quick rework. I'm about to finish the procd/busybox variants. On recompiling the Kernel I get a bunch of symbol questions, does the same happen to you?

if the symbols are unrelated to this PR, then it's probably because someone needs to add them to default config. Buildbots and the online CI have a bug and will always answer any prompt during kernel config with "enter", so will not notice if there are missing symbols from default config.

I sent a PR a few months ago and we found out about the buildbot bug https://github.com/openwrt/openwrt/pull/2948

aparcar commented on package/libs/libsepol/patches/010-Make-use-of-variables-when-defining-libdir-and-inclu.patch:
```diff
@@ -0,0 +1,26 @@
+Pkg-config definitions commonly make use of ${exec_prefix} and ${prefix}
```
Please remove this patch.

flyn-org replied:
Done.

aparcar commented on package/libs/libsepol/Makefile:
```diff
@@ -0,0 +1,65 @@
+define Build/InstallDev
+	$(INSTALL_DIR) $(1)/usr/include
+	$(CP) $(PKG_INSTALL_DIR)/usr/include/* $(1)/usr/include/
+	$(INSTALL_DIR) $(1)/usr/lib/pkgconfig
+	$(INSTALL_DATA) $(PKG_INSTALL_DIR)/usr/lib/pkgconfig/libsepol.pc $(1)/usr/lib/pkgconfig/
+	$(INSTALL_DIR) $(1)/usr/lib
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/* $(1)/usr/lib/
```
```suggestion
        $(INSTALL_DIR) $(1)/usr/lib
        $(CP) $(PKG_INSTALL_DIR)/usr/lib/* $(1)/usr/lib/
+       $(INSTALL_DIR) $(1)/usr/lib/pkgconfig
+       $(INSTALL_DATA) $(PKG_INSTALL_DIR)/usr/lib/pkgconfig/libsepol.pc $(1)/usr/lib/pkgconfig/
+       $(SED) 's,/usr/include,$$$${prefix}/include,g' $(1)/usr/lib/pkgconfig/libsepol.pc
+       $(SED) 's,/usr/lib,$$$${exec_prefix}/lib,g' $(1)/usr/lib/pkgconfig/libsepol.pc
```

flyn-org replied:
Done.

aparcar commented on package/libs/libselinux/Makefile:
```diff
@@ -0,0 +1,78 @@
+define Build/InstallDev
+	$(INSTALL_DIR) $(1)/usr/include
+	$(CP) $(PKG_INSTALL_DIR)/usr/include/* $(1)/usr/include/
+	$(INSTALL_DIR) $(1)/usr/lib/pkgconfig
+	$(INSTALL_DATA) $(PKG_INSTALL_DIR)/usr/lib/pkgconfig/libselinux.pc $(1)/usr/lib/pkgconfig/
+	$(INSTALL_DIR) $(1)/usr/lib
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/* $(1)/usr/lib/
```
```suggestion
        $(INSTALL_DIR) $(1)/usr/lib
        $(CP) $(PKG_INSTALL_DIR)/usr/lib/* $(1)/usr/lib/
+       $(INSTALL_DIR) $(1)/usr/lib/pkgconfig
+       $(INSTALL_DATA) $(PKG_INSTALL_DIR)/usr/lib/pkgconfig/libselinux.pc $(1)/usr/lib/pkgconfig
+       $(SED) 's,/usr/include,$$$${prefix}/include,g' $(1)/usr/lib/pkgconfig/libselinux.pc
+       $(SED) 's,/usr/lib,$$$${exec_prefix}/lib,g' $(1)/usr/lib/pkgconfig/libselinux.pc
```

flyn-org replied:
Done.

aparcar commented on package/libs/libsepol/Makefile:
```diff
@@ -0,0 +1,65 @@
+	PREFIX=$(STAGING_DIR_HOSTPKG) \
+	SHLIBDIR=$(STAGING_DIR_HOSTPKG)/lib
+
+$(eval $(call HostBuild))
```
Please move that line to the bottom, just above the other call.

flyn-org replied:
Done.

aparcar commented on package/libs/libselinux/Makefile:
```diff
@@ -0,0 +1,78 @@
+	PREFIX=$(STAGING_DIR_HOSTPKG) \
+	SHLIBDIR=$(STAGING_DIR_HOSTPKG)/lib
+
+$(eval $(call HostBuild))
```
Please move that line to the bottom, just above the other call.

flyn-org replied:
Done.

aparcar commented:
The commit `busybox: allow selection of SELinux applets` can be removed as `busybox` requires some additional refactoring which I'll take care of.

aparcar commented:
Here is the patch for both busybox and procd. https://github.com/openwrt/openwrt/pull/3303

dangowrt commented:
Did you intend to add host-builds for `libsepol` and and `pcre`? Right now they are referenced by libselinux Makefile but don't actually exist.

dangowrt reviewed (changes requested):
This triggers an armada of other symbols which are then undefined.
You should add eg.
```
config KERNEL_NETLABEL
    bool
```
and select that for `KERNEL_SECURITY_NETWORK`.
Not sure what to do about netfilter and conntrack modules, but to even get started we need the build to run through without any kernel oldconfig prompts.

Apart from that, I guess more filesystem xattr support will be needed in addition to EXT4 and SQUASHFS, at least F2FS, JFFS2 and UBIFS would be required to work with overlayfs with xattr on all platforms.

kofec commented:
I have below issue:
```
install -d -m0755 /home/amk082/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7620/iproute2-tiny/iproute2-5.7.0/ipkg-mipsel_24kc/ip-tiny/usr/libexec
install -m0755 /home/amk082/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7620/iproute2-tiny/iproute2-5.7.0/ip/ip /home/amk082/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7620/iproute2-tiny/iproute2-5.7.0/ipkg-mipsel_24kc/ip-tiny/usr/libexec/ip-tiny
find /home/amk082/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7620/iproute2-tiny/iproute2-5.7.0/ipkg-mipsel_24kc/ip-tiny -name 'CVS' -o -name '.svn' -o -name '.#*' -o -name '*~'| xargs -r rm -rf
Package ip-tiny is missing dependencies for the following libraries:
libselinux.so.1
make[3]: *** [/home/amk082/openwrt/bin/packages/mipsel_24kc/base/ip-tiny_5.7.0-2_mipsel_24kc.ipk] Error 1
make[3]: Leaving directory `/home/amk082/openwrt/package/network/utils/iproute2'
time: package/network/utils/iproute2/tiny/compile#0.17#0.24#0.37
make[2]: *** [package/network/utils/iproute2/compile] Error 2
make[2]: Leaving directory `/home/amk082/openwrt'
make[1]: *** [/home/amk082/openwrt/staging_dir/target-mipsel_24kc_musl/stamp/.package_compile] Error 2
make[1]: Leaving directory `/home/amk082/openwrt'
make: *** [world] Error 2
```

dangowrt commented:
g

> I have below issue:
> 
> ```
> install -d -m0755 /home/amk082/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7620/iproute2-tiny/iproute2-5.7.0/ipkg-mipsel_24kc/ip-tiny/usr/libexec
> install -m0755 /home/amk082/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7620/iproute2-tiny/iproute2-5.7.0/ip/ip /home/amk082/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7620/iproute2-tiny/iproute2-5.7.0/ipkg-mipsel_24kc/ip-tiny/usr/libexec/ip-tiny
> find /home/amk082/openwrt/build_dir/target-mipsel_24kc_musl/linux-ramips_mt7620/iproute2-tiny/iproute2-5.7.0/ipkg-mipsel_24kc/ip-tiny -name 'CVS' -o -name '.svn' -o -name '.#*' -o -name '*~'| xargs -r rm -rf
> Package ip-tiny is missing dependencies for the following libraries:
> libselinux.so.1
> make[3]: *** [/home/amk082/openwrt/bin/packages/mipsel_24kc/base/ip-tiny_5.7.0-2_mipsel_24kc.ipk] Error 1
> make[3]: Leaving directory `/home/amk082/openwrt/package/network/utils/iproute2'
> time: package/network/utils/iproute2/tiny/compile#0.17#0.24#0.37
> make[2]: *** [package/network/utils/iproute2/compile] Error 2
> make[2]: Leaving directory `/home/amk082/openwrt'
> make[1]: *** [/home/amk082/openwrt/staging_dir/target-mipsel_24kc_musl/stamp/.package_compile] Error 2
> make[1]: Leaving directory `/home/amk082/openwrt'
> make: *** [world] Error 2
> ```

Looks like iproute2 need to be patched to build without selinux (it's not an autotools project, so needs some work)

aparcar commented:
@flyn-org libselinux, libsepol are now part of openwrt.git, please rebase this PR.

flyn-org commented:
@aparcar, very cool! I updated the branch behind this PR.

adschm commented:
That's the magic of closing stuff ...

zuzia-dev commented:
I have this issue:

`Package libdevmapper is missing dependencies for the following libraries:
libselinux.so.1
libsepol.so.1
make[2]: *** [Makefile:115: /media/demo/DANE3/kompilacje/source-eb904-new-kernel-v8/openwrt/openwrt/bin/packages/mips_24kc/packages/libdevmapper_2.03.09-1_mips_24kc.ipk] Error 1
make[2]: Leaving directory '/media/demo/DANE3/kompilacje/source-eb904-new-kernel-v8/openwrt/openwrt/feeds/packages/utils/lvm2'
time: package/feeds/packages/lvm2/compile#75.19#21.44#104.97
make[1]: *** [package/Makefile:113: package/feeds/packages/lvm2/compile] Error 2
make[1]: Leaving directory '/media/demo/DANE3/kompilacje/source-eb904-new-kernel-v8/openwrt/openwrt'
make: *** [/media/demo/DANE3/kompilacje/source-eb904-new-kernel-v8/openwrt/openwrt/include/toplevel.mk:235: package/lvm2/compile] Błąd 2`

flyn-org commented:
> I have a similar problem:
> 
> `Package libdevmapper is missing dependencies for the following libraries: libselinux.so.1 libsepol.so.1 make[2]: *** [Makefile:115: /media/demo/DANE3/kompilacje/source-eb904-new-kernel-v8/openwrt/openwrt/bin/packages/mips_24kc/packages/libdevmapper_2.03.09-1_mips_24kc.ipk] Error 1 make[2]: Leaving directory '/media/demo/DANE3/kompilacje/source-eb904-new-kernel-v8/openwrt/openwrt/feeds/packages/utils/lvm2' time: package/feeds/packages/lvm2/compile#75.19#21.44#104.97 make[1]: *** [package/Makefile:113: package/feeds/packages/lvm2/compile] Error 2 make[1]: Leaving directory '/media/demo/DANE3/kompilacje/source-eb904-new-kernel-v8/openwrt/openwrt' make: *** [/media/demo/DANE3/kompilacje/source-eb904-new-kernel-v8/openwrt/openwrt/include/toplevel.mk:235: package/lvm2/compile] Błąd 2`

@aparcar, do you think this calls for a variant? Or should we just add a depends that is conditional on whether libselinux and libsepol are built?

aparcar commented:
I think the `CONFIGURE_ARGS` should contain `--disable-selinux` for now. I'm guessing a variant would be the right way to do, @flyn-org do you feel like implementing that?

@zuzia-dev if you post multiline pastes please use thre times `\`` so the line break work.

dangowrt commented:
> @aparcar, do you think this calls for a variant? Or should we just add a depends that is conditional on whether libselinux and libsepol are built?

As LVM2 is unlikely to be used on space constraint systems I wouldn't mind it always depending on libselinux and libsepol and pulling in 250kB of stuff which is useless on a non-SELinux system (but also doesn't do any harm).
Variants are worth doing for packages we are also using on very space-limited. For most things in packages which are useful to be built with libselinux on SELinux systems we should just always make it depend on libselinux (ie. just live with the fact that for some things packages additional 250kB will be wasted for dead code)

flyn-org commented:
> This triggers an armada of other symbols which are then undefined.
> You should add eg.
> 
> ```
> config KERNEL_NETLABEL
>     bool
> ```
> 
> and select that for `KERNEL_SECURITY_NETWORK`.
> Not sure what to do about netfilter and conntrack modules, but to even get started we need the build to run through without any kernel oldconfig prompts.
> 
> Apart from that, I guess more filesystem xattr support will be needed in addition to EXT4 and SQUASHFS, at least F2FS, JFFS2 and UBIFS would be required to work with overlayfs with xattr on all platforms.

@dangowrt, commit 3920bf0d should address the prompts regarding kernel features.

dangowrt commented:
LVM2 now always depends on libselinux https://github.com/openwrt/packages/commit/d8d06cfca262d97c7d3a3481c6a0f2cd3803c250

flyn-org commented:
I added a few more SELinux-related options to `config/Config-kernel.in`.

aparcar commented on include/image.mk:
```diff
@@ -234,13 +234,26 @@ endef
 
+ifeq ($(CONFIG_PACKAGE_refpolicy),y)
+define Image/mkfs/squashfs
+	echo "LD_LIBRARY_PATH=\$$LD_LIBRARY_PATH:$(STAGING_DIR_HOSTPKG)/lib $(STAGING_DIR_HOSTPKG)/sbin/setfiles -r $(call mkfs_target_dir,$(1)) $(call mkfs_target_dir,$(1))/etc/selinux/targeted/contexts/files/file_contexts $(call mkfs_target_dir,$(1))" > $@.fakeroot-script
```
We need the file `file_contexts` but can't have a Python package (`refpolicy`) in openwrt.git. Is it possible to have that single file as a package?

flyn-org replied:
I am confused. From what I see, the refpolicy package does not require Python, except on the build host. Should refpolicy move to openwrt.git. Am I missing something?

aparcar replied:
I talked to @dangowrt and he agrees that all three packages are required in openwrt.git. Please move them over and we're nearly done!

aparcar commented on package/utils/busybox/Makefile:
```diff
@@ -17,7 +17,7 @@ PKG_SOURCE_URL:=https://www.busybox.net/downloads \
 PKG_HASH:=d0f940a72f648943c1f2211e0e3117387c31d765137d92bd8284a3fb9752a998
 
-PKG_BUILD_DEPENDS:=BUSYBOX_CONFIG_PAM:libpam
+PKG_BUILD_DEPENDS:=BUSYBOX_CONFIG_PAM:libpam BUSYBOX_CONFIG_SELINUX:libselinux
```
Please remove this commit as it's covered in https://github.com/openwrt/openwrt/pull/3303

flyn-org replied:
I would like to wait until #3303 is merged. As before, I would like to keep this branch usable until it is merged.

aparcar replied:
Okay.

flyn-org replied:
@aparcar, I just removed the busybox commit, as merge request #3303 was merged.

aparcar replied:
Yea thanks for that!

tpetazzoni committed:
kernel: add options needed for SELinux

This adds a number of options to config/Config-kernel.in so that
packages related to SELinux support can enable the appropriate Linux
kernel support.

Signed-off-by: Thomas Petazzoni <thomas.petazzoni@bootlin.com>
[rebase; add ext4, F2FS, UBIFS, and JFFS2 support; add commit message]
Signed-off-by: W. Michael Petullo <mike@flyn.org>

flyn-org commented:
The most recent push adds the ability to enable security labels on F2FS, UBIFS, and JFFS2.

tpetazzoni committed:
policycoreutils: new package

Signed-off-by: Thomas Petazzoni <thomas.petazzoni@bootlin.com>
[update to 3.1, make use of Python 3, use ALTERNATIVES, and move to openwrt.git]
Signed-off-by: W. Michael Petullo <mike@flyn.org>

tpetazzoni committed:
checkpolicy: new package

Signed-off-by: Thomas Petazzoni <thomas.petazzoni@bootlin.com>
[update to 3.1, make use of Python 3, and move to openwrt.git]
Signed-off-by: W. Michael Petullo <mike@flyn.org>

tpetazzoni committed:
refpolicy: new package

Signed-off-by: Thomas Petazzoni <thomas.petazzoni@bootlin.com>
[update to 2.20200229, adjust Makefile, and move to openwrt.git]
Signed-off-by: W. Michael Petullo <mike@flyn.org>

flyn-org commented:
The policycoreutils, checkpolicy, and refpolicy packages are here. See also their removal from https://github.com/openwrt/packages/pull/10664.

aparcar commented:
> The policycoreutils, checkpolicy, and refpolicy packages are now here. See also their removal from openwrt/packages#10664.

Thank you, I give it another try.

aparcar commented:
New [Bootlog](http://sprunge.us/THHOW4)

Some SELinux messages
```
[   28.589884] audit: type=1400 audit(1598243556.768:3): avc:  denied  { write } for  pid=1007 comm="fw3" name="net" dev="proc" ino=757 scontext=system_u:system_r:initrc_t tcontext=system_u:system_r:initrc_t tclass=dir permissive=0
[   40.719390] audit: type=1400 audit(1598243568.896:4): avc:  denied  { write } for  pid=1253 comm="fw3" name="net" dev="proc" ino=1099 scontext=system_u:system_r:init_t tcontext=system_u:system_r:init_t tclass=dir permissive=0
```

Lastly looking at the labels:
```
root@OpenWrt:/# ls -Z
root@OpenWrt:/# chcon -R -t httpd_sys_content_t test
root@OpenWrt:/# ls -Z
system_u:object_r:bin_t          bin
system_u:object_r:tmpfs_t        dev
system_u:object_r:etc_t          etc
system_u:object_r:default_t      init
system_u:object_r:lib_t          lib
system_u:object_r:lib_t          lib64
system_u:object_r:mnt_t          mnt
system_u:object_r:unlabeled_t    overlay
system_u:object_r:proc_t         proc
system_u:object_r:root_t         rom
root:object_r:user_home_dir_t    root
system_u:object_r:bin_t          sbin
system_u:object_r:sysfs_t        sys
system_u:object_r:httpd_sys_content_t test
system_u:object_r:tmpfs_t        tmp
system_u:object_r:usr_t          usr
system_u:object_r:default_t      var
system_u:object_r:default_t      www
```

The file `/test` is manually added and therefore part of the overlay filesystem. I'm figuring therefore that labels work for both squashfs and overlay. If there's a better test for that, please let me know.

dangowrt commented on package/system/refpolicy/Makefile:
```diff
@@ -0,0 +1,78 @@
+  CATEGORY:=Base system
+  TITLE:=SELinux reference policy
+  URL:=http://selinuxproject.org/page/Main_Page
+  DEPENDS:=+@TARGET_ROOTFS_NEEDS_XATTR
```
I don't see that symbol `CONFIG_TARGET_ROOTFS_NEEDS_XATTR` defined or used anywhere. Am I missing something?

flyn-org replied:
I changed
```
DEPENDS:=+@TARGET_ROOTFS_NEEDS_XATTR
```
to
```
DEPENDS:=@KERNEL_SQUASHFS_XATTR||KERNEL_EXT4_FS_SECURITY||KERNEL_F2FS_FS_SECURITY||KERNEL_UBIFS_FS_SECURITY||KERNEL_JFFS2_FS_SECURITY
```
Please let me know if there is a better solution.

dangowrt commented on include/image.mk:
```diff
@@ -234,13 +234,26 @@ endef
 		-processors 1
 endef
 
+ifeq ($(CONFIG_PACKAGE_refpolicy),y)
```
This will not work for our binary builds having `CONFIG_ALL=y` and hence also `CONFIG_PACKAGE_refpolicy=y` always set.

flyn-org replied:
I changed this to
```
ifeq ($(CONFIG_KERNEL_SECURITY_SELINUX),y)
```
Does this solve the problem?

dangowrt replied:
I like Paul's suggestion of introducing an independent config symbol like you did in your most recent update of the PR.

flyn-org commented:
During my overnight build, I found that policycoreutils depended on PAM. The latest push makes this optional, depending on whether the busybox configuration included PAM support. The libpam package is in packages.git.

flyn-org commented:
@aparcar, I broke up the long echo command in the latest push. Presently running a full test build.

aparcar commented:
@flyn-org thanks for the update

What do you think about the following changes:

```diff
diff --git a/config/Config-build.in b/config/Config-build.in
index 499f84690f..e36bbd8372 100644
--- a/config/Config-build.in
+++ b/config/Config-build.in
@@ -328,4 +328,15 @@ menu "Global build settings"
                        bool "Full"
        endchoice
 
+       config CONFIG_TARGET_ROOTFS_SECURITY_LABELS
+               bool "Enable rootfs security labels"
+               select KERNEL_SQUASHFS_XATTR
+               select KERNEL_EXT4_FS_SECURITY
+               select KERNEL_F2FS_FS_SECURITY
+               select KERNEL_UBIFS_FS_SECURITY
+               select KERNEL_JFFS2_FS_SECURITY
+               select PACKAGE_refpolicy
+               help
+                 This option allows the usage of SELinux labels
+
 endmenu
diff --git a/include/image.mk b/include/image.mk
index afc336f946..0c9c0fc7a5 100644
--- a/include/image.mk
+++ b/include/image.mk
@@ -246,9 +246,12 @@ define Image/mkfs/squashfs-common
                -processors 1
 endef
 
-ifeq ($(CONFIG_PACKAGE_refpolicy),y)
+ifeq ($(CONFIG_TARGET_ROOTFS_SECURITY_LABELS),y)
 define Image/mkfs/squashfs
-       echo "LD_LIBRARY_PATH=\$$LD_LIBRARY_PATH:$(STAGING_DIR_HOSTPKG)/lib $(STAGING_DIR_HOSTPKG)/sbin/setfiles -r $(call mkfs_target_dir,$(1)) $(call mkfs_target_dir,$(1))/etc/selinux/targeted/contexts/files/file_contexts $(call mkfs_target_dir,$(1))" > $@.fakeroot-script
+       echo "LD_LIBRARY_PATH=\$$LD_LIBRARY_PATH:$(STAGING_DIR_HOSTPKG)/lib \
+               $(STAGING_DIR_HOSTPKG)/sbin/setfiles -r $(call mkfs_target_dir,$(1)) \
+               $(call mkfs_target_dir,$(1))/etc/selinux/targeted/contexts/files/file_contexts \
+               $(call mkfs_target_dir,$(1))" > $@.fakeroot-script
        echo "$(Image/mkfs/squashfs-common)" >> $@.fakeroot-script
        chmod +x $@.fakeroot-script
        $(STAGING_DIR_HOST)/bin/fakeroot $@.fakeroot-script
```

Long term we could introduce another option CONFIG_SELINUX which pulls in all those elements, like procd-selinux and more.

flyn-org commented:
@aparcar, I like your suggestion. The latest push follows your approach. I am doing a full build test now.

dangowrt commented on package/utils/policycoreutils/Makefile:
```diff
@@ -0,0 +1,80 @@
+MAKE_FLAGS += \
+	PAMH=$(CONFIG_BUSYBOX_CONFIG_PAM)
+
+define Package/policycoreutils/install
```
It'd be nice to break this up into individual packages, afaik also the optional libpam dependency then would only affect the `newrole` package (and then we could even have variants: `newrole` and `newrole-libpam`). See package [`util-linux`](https://github.com/openwrt/openwrt/blob/master/package/utils/util-linux/Makefile) for a good example.

However, I wouldn't mind to do that after the initial merge of that PR which is waiting too long already.

flyn-org replied:
I would greatly appreciate allowing this to happen post-merge.

dangowrt reviewed (changes requested):
Looks very good to me, still found issues with dependencies though.

dangowrt commented on config/Config-build.in:
```diff
@@ -328,4 +328,14 @@ menu "Global build settings"
 			bool "Full"
 	endchoice
 
+	config CONFIG_TARGET_ROOTFS_SECURITY_LABELS
```
This then results in `CONFIG_CONFIG_TARGET_ROOTFS_SECURITY_LABELS`. Please remove the redundant `CONFIG_` prefix.

flyn-org replied:
Fixed with latest push. Sorry, my test build failed about the same time you reported this.

dangowrt commented on config/Config-build.in:
```diff
@@ -328,4 +328,14 @@ menu "Global build settings"
+		select KERNEL_JFFS2_FS_SECURITY
+		select PACKAGE_refpolicy
+		help
+		  This option allows the usage of SELinux labels
```
`s/allows/enables` (especially as this effectively turns on labels in squashfs and installs refpolicy, that's more than 'allowing').

flyn-org replied:
Fixed with latest push.

flyn-org commented on package/system/refpolicy/Makefile:
```diff
@@ -0,0 +1,78 @@
+  CATEGORY:=Base system
+  TITLE:=SELinux reference policy
+  URL:=http://selinuxproject.org/page/Main_Page
+  DEPENDS:=+TARGET_ROOTFS_SECURITY_LABELS
```
I removed the DEPENDS from refpolicy's Makefile. Are you suggesting I add "depends on ... repolicy/host" to the "TARGET_ROOTFS_SECURITY_LABELS" block in "config/Config-build.in?" What is the syntax for that?

aparcar commented:
I'd say this is ready for testing. Enabling cergen for wolfssl doesn't seem to have any size impact, I'm guessing it's selected by one of the other options as well. The compiled px5g-wolfssl tool is 5kB in size.

aparcar commented on package/utils/policycoreutils/Makefile:
```diff
@@ -0,0 +1,80 @@
+HOST_BUILD_DEPENDS:=libsemanage/host
+PKG_BUILD_DEPENDS:=BUSYBOX_CONFIG_PAM:libpam
+
+PKG_MAINTAINER:=Thomas Petazzoni <thomas.petazzoni@bootlin.com>
```
```suggestion
PKG_MAINTAINER:=Thomas Petazzoni <thomas.petazzoni@bootlin.com>
PKG_CPE_ID:=cpe:/a:selinuxproject:policycoreutils
PKG_LICENSE:=GPL-2.0-or-later
PKG_LICENSE_FILES:=COPYING
```

flyn-org replied:
Done!

aparcar commented on package/utils/checkpolicy/Makefile:
```diff
@@ -0,0 +1,48 @@
+PKG_INSTALL:=1
+HOST_BUILD_DEPENDS:=libselinux/host
+
+PKG_MAINTAINER:=Thomas Petazzoni <thomas.petazzoni@bootlin.com>
```
```suggestion
PKG_MAINTAINER:=Thomas Petazzoni <thomas.petazzoni@bootlin.com>
PKG_CPE_ID:=cpe:/a:selinuxproject:checkpolicy
PKG_LICENSE:=GPL-2.0-or-later
PKG_LICENSE_FILES:=COPYING
```

flyn-org replied:
Done!

aparcar commented on package/system/refpolicy/Makefile:
```diff
@@ -0,0 +1,77 @@
+PKG_INSTALL:=1
+PKG_BUILD_DEPENDS:=checkpolicy/host policycoreutils/host
+
+PKG_MAINTAINER:=Thomas Petazzoni <thomas.petazzoni@bootlin.com>
```
```suggestion
PKG_MAINTAINER:=Thomas Petazzoni <thomas.petazzoni@bootlin.com>
PKG_CPE_ID:=cpe:/a:tresys:refpolicy
PKG_LICENSE:=GPL-2.0-or-later
PKG_LICENSE_FILES:=COPYING
```
The CPE here is somewhat optional as `refpolicy` moved over to SELinuxProject. I'll ping them upstream.

aparcar replied:
https://github.com/SELinuxProject/refpolicy/issues/306

flyn-org replied:
Done!

aparcar commented:
Please update the metadata of added packages, everything else looks good to me.

dangowrt commented:
Pulled. Thanks everyone for the hard work!

aparcar commented:
🥂

flyn-org commented:
Thank you everyone!

dhewg commented:
Looks like libsemanage should be part of openwrt.git too?
```
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a dependency on 'libsemanage', which does not exist
WARNING: Makefile 'package/utils/policycoreutils/Makefile' has a build dependency on 'libsemanage/host', which does not exist
```

flyn-org commented:
Looks like @dangowrt already pulled in the libsemanage dependency.

aparcar commented:
https://patchwork.ozlabs.org/project/openwrt/patch/20200902003245.3315830-1-mail@aparcar.org/ allows tu have selinux enabled without any cmdline modification

aparcar commented:
@tpetazzoni @flyn-org procd requires some more work to set the correct context for running services. Could you add that feature to complete the SELinux setup?

aparcar commented:
I wrote a short blog post on how to enable SELinux in it's current state:
https://aparcar.gitlab.io/running-openwrt-with-selinux/
